Make Quickstart session page visible when logged out

niranjan-uma-shankar opened the pull request:
#### Changes proposed in this Pull Request

* Shows the Quickstart session page even to a logged out user. 

Navigating to `http://calypso.localhost:3000/checkout/SITE_SLUG/add-quickstart-session` while logged out will show the Quickstart Session offer page with a `Get Started` button. With Calypso live, you can instead navigate to `https://hash-7387affcb976116ea8f0a9d85df15ae16b5ea495.calypso.live/checkout/SITE_SLUG/add-quickstart-session`

![calypso localhost_3000_checkout_iop280518100 wordpress com_add-quickstart-session](https://user-images.githubusercontent.com/1269602/59347690-e1578000-8d32-11e9-9950-1d26fc7d3740.png)


Clicking `Get Started` should take you to the Log In page. After successful signing in, you should be redirected to the Order Summary page with the Support Session already added to cart. Check this [recording](https://media.giphy.com/media/VEc4Uunvcv6K4wO9xH/giphy.mp4).

Please note that for the Order Summary page to be shown, the SITE_SLUG should be a valid slug for the account that you are signing in. 



#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Scenario 1**

* Log out of WP.com and navigate to `http://calypso.localhost:3000/checkout/SITE_SLUG/add-quickstart-session` OR `https://hash-7387affcb976116ea8f0a9d85df15ae16b5ea495.calypso.live./checkout/SITE_SLUG/add-quickstart-session`. The SITE_SLUG should be a valid slug associated with the account that you plan to sign in with.
* Verify from the previous step that you can see the Quickstart session offer page with a "Get Started"
  button.
* Click the "Get Started" button, and verify that you are taken to a Log In page.
* Sign in with your account, and verify that you are taken to the Order Summary page with the support session item already added to cart. 
* Complete payment and verify that the purchase succeeds.

**Scenario 2**

* While logged in to WP.com, verify that `http://calypso.localhost:3000/checkout/SITE_SLUG/add-quickstart-session` opens the Quickstart session offer page with Accept and Decline buttons as shown below:

![calypso localhost_3000_checkout_iop280518100 wordpress com_add-quickstart-session (1)](https://user-images.githubusercontent.com/1269602/59348913-000b4600-8d36-11e9-9942-a639660274c9.png)

Complete the checkout process and verify that the purchase is successful

**Scenario 3**

Verify _Scenario 1_ on a mobile device. Verify that alignments are proper and nothing stands out as abnormal.

**Scenario 4**

While logged out, visiting any other /checkout route should redirect to the log in page. So try the following routes:

http://calypso.localhost:3000/checkout
http://calypso.localhost:3000/checkout/SITE_SLUG/add-support-session
http://calypso.localhost:3000/checkout/SITE_SLUG/add-support-session
http://calypso.localhost:3000/checkout/thank-you/no-site

While logged in, verify that all the above routes redirect to their respective destinations. Note that while logged in, `http://calypso.localhost:3000/checkout` without a plan or product will redirect to 
/plans.

**Scenario 5**

Try accessing the logged out offer page from different locations (USA, UK, Japan etc) using a VPN service, and verify that the currency denomination changes for these different locations.

niranjan-uma-shankar committed:
First commit

niranjan-uma-shankar committed:
testing

niranjan-uma-shankar committed:
Add logged out route for the quickstart session page; redirect any other checkout route to the login page

niranjan-uma-shankar committed:
Show a 'Get Started' button while logged out; Handle click event on the Get Started button by redirecting it to the login page, and handle post login redirect to the checkout page; CSS changes for these

niranjan-uma-shankar commented on client/my-sites/checkout/concierge-quickstart-session/index.jsx:
```diff
@@ -62,6 +63,14 @@ export class ConciergeQuickstartSession extends React.Component {
+		// If the param value is 'add', then we will add the concierge session item to cart and proceed to checkout,
+		// without showing any UI. We use this method when a logged out user clicks 'Get Started' on the concierge
+		// offer page.
+		if ( 'add' === receiptId ) {
```
`receiptId ` is an unfortunate name but I am following up with another PR that cleans up this function and will do away with the need for a receiptId, so that should take care of this.

niranjan-uma-shankar committed:
Remove commented out text

niranjan-uma-shankar committed:
Resolving a few warnings/errors

1. selectedSiteId is now no longer a required propType, given that the quickstart offer page can be access while logged out 2. receiptId is no longer ju
t a number in the quickstart offer page 3. If the receiptId val is `add`, then skip the UI and directly take to checkout. This needs to be handled in th
e component life cycle methods, instead of in render() 4. Move redirectLoggedOut to mapStateToProps to avoid action undefined error in wpcom-api-middleware.js

niranjan-uma-shankar committed:
selectedSite is no longer a required prop, given that we can access this page while logged out

niranjan-uma-shankar committed:
Renaming Support Session to Quick Start

niranjan-uma-shankar commented on client/sections.js:
```diff
@@ -251,6 +251,7 @@ const sections = [
 		module: 'my-sites/checkout',
 		secondary: true,
 		group: 'sites',
+		enableLoggedOut: true,
```
I've verified that all /checkout routes redirect to log in, except for the quick start route. The condition for this is [here](https://github.com/Automattic/wp-calypso/pull/33749/files#diff-137f56ac9700496d02aebf4e8ad62b04R38). Also pinging team-calypso to make sure there are no gotchas.

niranjan-uma-shankar replied:
I heard back that there are no issues with this approach in C03DY3RJ5/p1560791734013400-slack.

niranjan-uma-shankar commented:
Thanks for the review @travisw ! I've now fixed the errors in [cb6c868](https://github.com/Automattic/wp-calypso/pull/33749/commits/cb6c86822a52d0435d0bced80f1323393f4c58a9) and [b51af52](https://github.com/Automattic/wp-calypso/pull/33749/commits/b51af5266c946e6419836ace446f5a7a4109e4c8).

----------------------------------------------------------------------------------------------------

Regarding this error you observed:


![](https://user-images.githubusercontent.com/690843/59473021-686d2c80-8df5-11e9-9cc3-ffcdf419ed6f.png)


I see this error in the console even when I just log in at http://calypso.localhost:3000/log-in. It appears to be a security policy to prevent XSS and likely shows up because we are not running on *.wordpress.com while on localhost. It doesn't seem to be introduced by this PR, so I think we can ignore it.

----------------------------------------------------------------------------------------------------

While logged out and on the quick start offer page, I see this message:

<img width="1671" alt="Screenshot 2019-06-17 at 10 27 48 PM" src="https://user-images.githubusercontent.com/1269602/59622436-2fafb900-914f-11e9-8393-5fcd6f8f29c8.png">

I'm not clear where this originates from. @Automattic/team-calypso would you be knowing about this? Have you seen this before, or know what could be causing it?

niranjan-uma-shankar committed:
Only set ID when it's not null. This prevents an 'Uncaught AuthorizationRequiredError' when viewing the quick start offer page while logged out

niranjan-uma-shankar commented:
The uncaught Authentication error notice has been fixed in [5ff14ec](https://github.com/Automattic/wp-calypso/pull/33749/commits/5ff14ec7c82fe6eb26c9a257c9222803e79d568c)

dzver commented on client/lib/cart/store/index.js:
```diff
@@ -241,6 +241,7 @@ function createListener( store, selector, callback ) {
 getReduxStore().then( store => {
-	CartStore.setSelectedSiteId( getSelectedSiteId( store.getState() ) );
+	const selectedSiteId = getSelectedSiteId( store.getState() );
+	selectedSiteId && CartStore.setSelectedSiteId( selectedSiteId );
```
What does this line of code do?

niranjan-uma-shankar replied:
@dzver Answered in this [comment](https://github.com/Automattic/wp-calypso/pull/33749#discussion_r295362510).

dzver reviewed (commented):
It works fine. There's one line of code that I'm not sure is fine.

There's one issue I think is missing. If I log in with another user, not the one that owns the site slug, it hangs in some incomplete state. It'd be better to have a redirect of some sort.

jsnajdr commented on client/my-sites/checkout/index.js:
```diff
@@ -16,13 +16,31 @@ import {
 	SiftScience.recordUser();
 
+	const user = userFactory();
+	const isLoggedOut = ! user.get();
```
It's better to use the `isUserLoggedIn` Redux selector here rather than the legacy `lib/user` module (we're trying to get rid of it)

niranjan-uma-shankar replied:
This file `client/my-sites/checkout/index.js` does not have a redux connect - do you suggest introducing one for this?

jsnajdr replied:
Hmm, this is not a React component, so `connect` cannot be used and there is no other way to access the Redux store from the section initialization function 🙁

There's another section that has a mixture of logged-in and logged-out views: [Themes](https://github.com/Automattic/wp-calypso/blob/c50e1d7c7496011e51fa5c114fdc4b9f2c345f3e/client/my-sites/themes/index.web.js#L22-L23). And it uses the `lib/user` module, too. It seems like the only solution right now. We might want to invent some better framework solution later, but that's not for this PR. You can dismiss my earlier comment and leave `lib/user` in there 🙂

jsnajdr commented on client/my-sites/checkout/index.js:
```diff
@@ -16,13 +16,31 @@ import {
+			conciergeQuickstartSession,
+			makeLayout,
+			clientRender
+		);
```
If the paths with `/add` and `/:receiptId` require login and only the root path supports logged-out view, then this handler is incorrect: navigating directly to one of these two URLs won't redirect to login.

niranjan-uma-shankar replied:
Fixed in [9105ae3](https://github.com/Automattic/wp-calypso/pull/33749/commits/9105ae36892a7c9b5d76260f39c9fecc8579f272).

niranjan-uma-shankar committed:
When 'Get Started' button is clicked, it no longer calls this.props.redirectLoggedOut() but instead redirects to /add. The /add route's handler will redirect to the login page if needed; modifies the route handler so that /checkout/:site/add-quickstart-session/add will redirect to login page

niranjan-uma-shankar committed:
Use getSelectedSiteId selector instead of getSelectedSite

niranjan-uma-shankar commented on client/my-sites/checkout/controller.jsx:
```diff
@@ -150,16 +150,15 @@ export function conciergeQuickstartSession( context, next ) {
-	if ( ! selectedSite ) {
-		return null;
-	}
+	const selectedSiteId = selectedSite && selectedSite.ID;
```
Fixed in [733ff2b](https://github.com/Automattic/wp-calypso/pull/33749/commits/733ff2b6550d6c7254137931d230f40a22661af0).

niranjan-uma-shankar commented on client/my-sites/checkout/concierge-quickstart-session/index.jsx:
```diff
@@ -308,30 +318,52 @@ export class ConciergeQuickstartSession extends React.Component {
+	 */
+	handleRedirect = () => {
+		const redirectPath = this.props.path + '/add';
+		this.props.redirectLoggedOut( { path: redirectPath, store: this.props.reduxStore } );
```
Fixed in [9105ae3](https://github.com/Automattic/wp-calypso/pull/33749/commits/9105ae36892a7c9b5d76260f39c9fecc8579f272).

jsnajdr commented on client/my-sites/checkout/concierge-quickstart-session/index.jsx:
```diff
@@ -45,10 +45,19 @@ import './style.scss';
+		// offer page.
+		if ( 'add' === this.props.receiptId ) {
+			this.handleClickAccept();
+		}
```
If you don't want to render any UI at all for the `../add` route, you can add the product to cart and redirect to checkout in the page.js handler (`conciergeQuickstartSession` function in `my-sites/checkout/controller.jsx`). And render the `ConciergeQuickstartSession` component only if there's a numeric `receiptId` and an actual UI to render.

niranjan-uma-shankar replied:
Good point. In [9849222](https://github.com/Automattic/wp-calypso/pull/33749/commits/98492222ecd61d3624a0558dcc8394047785e873), I've handled this in a slightly different (and hopefully better) way by removing the `/add` route altogether. Instead, a redirect happens to `/checkout/SITE_SLUG/quickstart-session`, so that the controller for the /checkout route( `function checkout ()` in `my-sites/checkout/controller.jsx`) will call the `Checkout` component which can handle addition of new products, similar to routes such as `/checkout/SITE_SLUG/business` and `/checkout/SITE_SLUG/premium`. 

One slight difference introduced by this method is that after clicking `Yes, I want a WordPress expert` button OR `Get Started` button(if logged out) in the [offer page](https://user-images.githubusercontent.com/1269602/59348913-000b4600-8d36-11e9-9942-a639660274c9.png), it will redirect to `/checkout/SITE_SLUG/quickstart-session` instead of the earlier `/checkout/SITE_SLUG`. I'm not sure if this is a problem, but just making a note here.

jsnajdr commented on client/my-sites/checkout/controller.jsx:
```diff
@@ -127,15 +127,15 @@ export function conciergeSessionNudge( context, next ) {
-	if ( ! selectedSite ) {
+	if ( ! selectedSiteId ) {
 		return null;
 	}
```
This check should not be neccessary. The route that is handled here always has a `:site` parameter, there is no support for "All Sites" view.

And if the specified `:site` is invalid (e.g., site doesn't exist), the `siteSelection` page.js handler will detect that and won't pass control to the next handler.

The `ConciergeSessionNudge` could fetch the `selectedSiteId` from Redux in its `connect` call. The only prop that needs to come outside Redux is `receiptId`:
```jsx
<ConciergeSessionNudge receiptId={ Number( receiptId ) } />
```

niranjan-uma-shankar replied:
Thanks, taken care of in [9849222](https://github.com/Automattic/wp-calypso/pull/33749/commits/98492222ecd61d3624a0558dcc8394047785e873). `ConciergeSessionNudge` component also gets a `siteSlugParam ` prop, which is used to redirect a logged out user(where we have no site ID) to `/checkout/SITE_SLUG/quickstart-session` after clicking the `Get Started` button [here](https://user-images.githubusercontent.com/1269602/59347690-e1578000-8d32-11e9-9950-1d26fc7d3740.png).

jsnajdr commented on client/my-sites/checkout/controller.jsx:
```diff
@@ -149,17 +149,15 @@ export function conciergeQuickstartSession( context, next ) {
-
-	if ( ! selectedSite ) {
-		return null;
-	}
```
The same case as with `ConciergeSessionNudge`: the check is not needed and `selectedSiteId` can be connected in the component itself.

jsnajdr commented on client/my-sites/checkout/controller.jsx:
```diff
@@ -149,17 +149,15 @@ export function conciergeQuickstartSession( context, next ) {
+				receiptId={ receiptId }
+				selectedSiteId={ selectedSiteId }
+				path={ context.path }
+				reduxStore={ context.store }
```
The `reduxStore` prop is no longer needed after the component stopped using `redirectIfLoggedOut`.

niranjan-uma-shankar replied:
Thanks, taken care of in [9849222](https://github.com/Automattic/wp-calypso/pull/33749/commits/98492222ecd61d3624a0558dcc8394047785e873).

niranjan-uma-shankar committed:
Make classnames show the full name instead of concatenating with &, for better searchability

niranjan-uma-shankar committed:
Do not poll the server to synchronize the cart if the user is not logged in, otherwise CartSynchronizer.prototype._performRequest throws an Authentication error in the console

niranjan-uma-shankar committed:
Renamed the add-quickstart-session slug to offer-quickstart-session because this route only shows the offer page for the Quick Start session and does not add anything to the cart.

niranjan-uma-shankar committed:
Use <Checkout> component to handle cart

The following changes are made in this revision:
1. handleClickAccept in ConciergeQuickstartSession component will no longer add item to cart, but instead will redirect to /checkout so that the Checkout component can handle adding the new concierge product to cart.
2. handleRedirect in ConciergeQuickstartSession component will no longer redirect to /add-quickstart-session/add, but instead will redirect to /checkout/SITE_SLUG/quickstart-session route so that the Checkout component can handle adding the new concierge product to cart The URL format /checkout/SITE_SLUG/quickstart-session is keeping in line with other formats such as /checkout/SITE_SLUG/business. Also, reusing the Checkout component to handle adding new products to cart is much better than duplicating the work in ConciergeQuickstartSession product
3. Redirecting /checkout/SITE_SLUG/quickstart-session to the Login page while visiting it logged out is automatically handled
4. Above changes are made to the concierge session component too
5. Removed unnecessary state assignments and other props

niranjan-uma-shankar commented on client/my-sites/checkout/concierge-quickstart-session/style.scss:
```diff
@@ -140,4 +140,8 @@
+
+	&__footer &__get-started-button {
+		width: 100%;
+	}
```
Fixed in [3b56859](https://github.com/Automattic/wp-calypso/pull/33749/commits/3b56859a76d75863d0a692309e914b374eeb24ec)

niranjan-uma-shankar commented on client/my-sites/checkout/checkout/index.jsx:
```diff
@@ -433,7 +441,7 @@ export class Checkout extends React.Component {
 			// Show them the concierge session upsell page
 			if ( 'offer' === abtest( 'conciergeUpsellDial' ) ) {
-				return `/checkout/${ selectedSiteSlug }/add-quickstart-session/${ receiptId }`;
+				return `/checkout/${ selectedSiteSlug }/offer-quickstart-session/${ receiptId }`;
```
Just to explain, this is due to the change introduced in [2a36a68 ](https://github.com/Automattic/wp-calypso/pull/33749/commits/2a36a68c7c90f10c713112d98cd49e88ce2a6c99)

niranjan-uma-shankar commented on client/lib/cart/store/index.js:
```diff
@@ -241,6 +241,7 @@ function createListener( store, selector, callback ) {
 getReduxStore().then( store => {
-	CartStore.setSelectedSiteId( getSelectedSiteId( store.getState() ) );
+	const selectedSiteId = getSelectedSiteId( store.getState() );
+	selectedSiteId && CartStore.setSelectedSiteId( selectedSiteId );
```
The reason I did this change was because while logged out, opening the Quick Start offer page showed the following error in the console:

![](https://user-images.githubusercontent.com/1269602/59622436-2fafb900-914f-11e9-8393-5fcd6f8f29c8.png)

> The original code set the cart site ID to null if some "All Sites" view is active. That disables the cart synchronization for that site.

In the following line, we still subscribe to the the redux store so wouldn't any null values still get set? Nevertheless, in [e1f863c](https://github.com/Automattic/wp-calypso/pull/33749/commits/e1f863c94e6e7812e531161b7dfcf2068dbdff82), I have removed this change. I have introduced a change wherein `setSelectedSiteId` proceeds only if the user is logged in, so that while logged out we don't poll the server for syncing the cart. Would this work, or is there something I am overlooking?

niranjan-uma-shankar committed:
Rename quickstart-session slug to concierge-session

niranjan-uma-shankar committed:
Keep the offer page track events in the component, instead of at the Checkout component. This helps avoid false positives when the add concierge product to cart link is directly called

niranjan-uma-shankar commented on client/my-sites/checkout/concierge-session-nudge/index.jsx:
```diff
@@ -333,38 +331,35 @@ export class ConciergeSessionNudge extends React.Component {
-		addItem( conciergeSessionItem() );
-
-		page( `/checkout/${ siteSlug }` );
+		page( `/checkout/${ this.props.siteSlug }/quickstart-session` );
```
This is a good point. I have implemented this in [8ba81a6](https://github.com/Automattic/wp-calypso/pull/33749/commits/8ba81a6719dad784e97f993d70fd7f61dd7b9c7c).

niranjan-uma-shankar commented on client/my-sites/checkout/checkout/index.jsx:
```diff
@@ -287,6 +288,13 @@ export class Checkout extends React.Component {
 			cartItem = domainMapping( { domain: cartMeta } );
 		}
 
+		if ( startsWith( this.props.product, 'quickstart-session' ) ) {
```
This is a good point. I have implemented this in [8ba81a6](https://github.com/Automattic/wp-calypso/pull/33749/commits/8ba81a6719dad784e97f993d70fd7f61dd7b9c7c).

I have used `concierge-session` in the URL, as you suggested in [this comment](https://github.com/Automattic/wp-calypso/pull/33749#discussion_r296068147). If you feel it's better to use support-session, I can quickly change that too.

niranjan-uma-shankar commented on client/my-sites/checkout/concierge-quickstart-session/index.jsx:
```diff
@@ -361,38 +372,36 @@ export class ConciergeQuickstartSession extends React.Component {
 	handleClickAccept = () => {
-		const { siteSlug, trackUpsellButtonClick } = this.props;
-
-		trackUpsellButtonClick( 'accept' );
```
I have implemented this in [98e7cf5](https://github.com/Automattic/wp-calypso/pull/33749/commits/98e7cf52d5be5495a3cbc58090bc419896580863). I've also added a tracking for the get started button.

niranjan-uma-shankar commented on client/my-sites/checkout/controller.jsx:
```diff
@@ -121,45 +121,33 @@ export function gsuiteNudge( context, next ) {
 }
 
 export function conciergeSessionNudge( context, next ) {
```
I will take this as a follow up in a subsequent PR.
